### ISSUE TITLE
Images pushed to GCR can't be pulled by docker

### ISSUE THREAD

**Issue by vugardzhamalov** [containers/skopeo#610]
Hi there,

It seems that I can push my image to private google container registry just fine... but Docker can't pull it afterwards... Am I doing something wrong or is this something expected?

I am pushing it like so:
skopeo --insecure-policy copy containers-storage:<image_id> docker://gcr.io/<project_id>/<image_name>:<image_tag>

Here is example of docker pull with error message:
docker pull gcr.io/<project_id>/<image_name>:<image_tag>

Error response from daemon: manifest for gcr.io/<project_id>/<image_name>:<image_tag> not found

**Comment by mtrmac** [comment]
Thanks for your report. This is not a known failure.

Does `skopeo --insecure-policy copy docker://gcr.io/<project_id>/<image_name>:<image_tag> dir:$someTempDir` work?

Wildly guessing, after pushing with `… copy --format=v2s2 …`, is the image visible to Docker?

Also, maybe the full output of `skopeo --insecure-policy --debug copy containers-storage:<image_id> docker://gcr.io/<project_id>/<image_name>:<image_tag>` (if at all possible, with all name components unredacted) might be helpful.

**Comment by vugardzhamalov** [comment]
Thank you @mtrmac,

I think I've managed to localize the issue - it is primarily between the keyboard and the chair on this side of the screen :-)

I was copying my local Buildah registry images to remote (GCR) registry (docker only?) ... It seems that I've missed important step - there is a need to export my local Buildah images to my local Docker image registry(?) beforehand... Hence the use of 'containers-storage:' in my original question as opposed to 'docker-daemon:'

So following steps worked for me, I've copied local buildah image to local docker registry:
`skopeo --insecure-policy copy containers-storage:<builda_image_id> docker-daemon:<docker_image_name>:<docker_image_tag>`

After that I've confirmed both Skopeo copy and Docker tag/push both working as expected and GCE kubectl deployment behaving as expected for any of the images stored on GCR, my Skopeo example:
`skopeo --insecure-policy --debug copy --format=v2s2 docker-daemon:<docker_image_name>:<docker_image_tag> docker://gcr.io/<project_id>/<destination_image_name>:<destination_image_tag>`

This ticket is probably not the best place to ask further questions and I will close it  - but I was wondering if it means that despite the fact that I can push my Buildah images to GCR there is no way for me deploy them to GCE? Any info will be much appreciated! Thank you!

**Comment by vugardzhamalov** [comment]
And I suspect that following [statement](https://cloud.google.com/kubernetes-engine/docs/tutorials/hello-app#step_1_build_the_container_image) answers my question:

> GKE accepts Docker images as the application deployment format. To build a Docker image, you need to have an application and a Dockerfile.

Still will be interesting to know if someone figured out how to create GKE deployment based on Buildah images as opposed to Docker.

**Comment by vugardzhamalov** [comment]
OK I suspect that [CBI](https://github.com/containerbuilding/cbi) could be a potential answer to my question. Closing!

**Comment by mtrmac** [comment]
> I think I've managed to localize the issue - it is primarily between the keyboard and the chair on this side of the screen :-)

I don’t think so…

> I was copying my local Buildah registry images to remote (GCR) registry (docker only?) ... It seems that I've missed important step - there is a need to export my local Buildah images to my local Docker image registry(?) beforehand...

(`docker-daemon:` means the local, Docker-the-daemon-managed, storage (which contains exploded filesystems); that’s different from `docker://`, which accesses a Docker registry (which contains compressed tarballs). Obviously we don’t want to _require_ a Docker daemon in order for Buildah to work.)

> So following steps worked for me, I've copied local buildah image to local docker registry:
> `skopeo --insecure-policy copy containers-storage:<builda_image_id> docker-daemon:<docker_image_name>:<docker_image_tag>`
> 
> After that I've confirmed both Skopeo copy and Docker tag/push both working as expected and GCE kubectl deployment behaving as expected for any of the images stored on GCR, my Skopeo example:
> `skopeo --insecure-policy --debug copy --format=v2s2 docker-daemon:<docker_image_name>:<docker_image_tag> docker://gcr.io/<project_id>/<destination_image_name>:<destination_image_tag>`

The intermediate step really should not be required; I suspect it’s actually the `--format=v2s2` that makes the difference.

Could you recheck whether
```
skopeo --insecure-policy copy --format=v2s2 containers-storage:<builda_image_id> docker://gcr.io/<project_id>/<destination_image_name>:<destination_image_tag>
```

works as expected, please?

Or, more directly, something like
```
buildah push --format=v2s2 [source] destination
```

Either way, the `--format=v2s2` option _should not be required_; the code should autodetect whether it is necessary, and it’s a bug that it does; but I’d like to first confirm that this is indeed the issue.


> This ticket is probably not the best place to ask further questions and I will close it - but I was wondering if it means that despite the fact that I can push my Buildah images to GCR there is no way for me deploy them to GCE? Any info will be much appreciated! Thank you!

I’m afraid I have no experience in this area yet; hopefully someone else in this project does.

**Comment by vugardzhamalov** [comment]
Thank you @mtrmac,

Interesting... I'll give this another go in a day or two. You'll hear from me as soon as I know more.

**Comment by vugardzhamalov** [comment]
Hi @mtrmac,

I think my new findings still confirm that this is pretty much a user error rather then anything else really...
For the sake of completeness here is my minimal script example that I build the image with, the only interesting thing about it - 'buildah commit' declaration with '--format' switch:

`#!/usr/bin/env bash
set -o errexit

# DECLARE CONTAINER VARIABLE
container01=$(buildah from alpine:3.8)

# COMMIT CONTAINER
buildah commit --format docker $container01 skopeo_gcr_test:1.0`

Now I execute the script and end up with 'skopeo_gcr_test' image. Because I've explicitly set format to 'docker' I can execute skopeo without declaring '--format' switch at all, but I've checked with explicit declaration as well - both  work as expected when I push it to GCR. In both cases image reported by GCR as image type: 'Docker Manifest, Schema 2'

Now if I execute 'buildah commit' without explicitly declaring '--format' switch things behave in a slightly different manner since it defaults to 'oci' format.

In case if I copy image with skopeo without specifying '--format' switch explicitly as you've instructed me earlier it will copy image as is. This time GCR will report image type as 'OCI Manifest'. At the same time when I call for skopeo with '--format=v2s2' switch it will convert my 'oci' image sort of on the fly and I will end up with the  'Docker Manifest, Schema 2' image type in my GCR. And if you'd ask me it looks as if it all works as expected...

Since by default GCP Kubernetes doesn't expect anything but Docker image deployment manifest will end up failing in case if you are referring to image type of 'oci' - but this is a completely different story I guess.

The only question I have now. Would you know how to fix this warning message I get into when I execute skopeo copy? Thank you!
`WARN[0000] Not using native diff for overlay, this may cause degraded performance for building images: failed to mount overlay: invalid argument`

**Comment by vugardzhamalov** [comment]
Hi @mtrmac,

Well it seems that I have to take some of my earlier statements back. I've just discovered that even if 'buildah commit' was issued with '--format docker' switch one still needs to use '--format=v2s2' switch with 'skopy copy'. Unless '--format=v2s2' switch was explicitly declared any attempt to Deploy container from the image stored on GCR will fail... I don't know if this is expected. I've checked manifests for both GCR images and they are identical but you still cannot deploy any containers if image was skopeo copied without '--format=v2s2' switch.
